File: ext/hallon/session.c
#include "common.h"
#include "callbacks.h" /* hn_session_fire */

/*
  Useful macros :D
*/
#define SESSPTR_OF(obj) *((sp_session**) DATA_OF(obj)->spotify_ptr)
#define DATA_OF(obj) Data_Fetch_Struct(obj, hn_spotify_data_t)

/*
  Prototypes
*/
static VALUE sp_session_process_events_nogvl(void *);
static VALUE sp_session_login_nogvl(void *);
static VALUE sp_session_logout_nogvl(void *);

/*
  @overload initialize(appkey, options = {}, &block)

    Creates a new Spotify session with the given parameters using `sp_session_create`.
  
    @example
       session = Hallon::Session.instance(appkey, :settings_path => "tmp") do
         on(:logged_in) do |error|
           puts "We logged in successfully. Lets bail!"
           exit
         end
       end
  
    @note Until `libspotify` allows you to create more than one session, you must
          use {Hallon::Session.instance} instead of this method.
  
    @param [#to_s] appkey your `libspotify` application key.
    @param [Hash] options additional options
    @option options [String] :user_agent ("Hallon") User-Agent to use (length < 256)
    @option options [String] :settings_path ("tmp") where to save settings and user-specific cache
    @option options [String] :cache_path ("") where to save cache files (set to "" to disable)
    @option options [Bool]   :load_playlists (true) load playlists into RAM on startup
    @option options [Bool]   :compress_playlists (true) compress local copies of playlists
    @option options [Bool]   :cache_playlist_metadata (true) cache metadata for playlists locally
    @yield allows you to define handlers for events (see {Base#on})
    @raise [ArgumentError] if options[:user_agent] is more than 255 characters long
    @raise [Hallon::Error] if `sp_session_create` fails
    @see http://developer.spotify.com/en/libspotify/docs/structsp__session__config.html
*/
static VALUE cSession_initialize(int argc, VALUE *argv, VALUE self)
{
  VALUE appkey, options;
  hn_spotify_data_t *session_data = DATA_OF(self);
  rb_call_super(0, NULL); // IMPORTANT, see Hallon::Base
  
  /* handle options */
  rb_scan_args(argc, argv, "12", &appkey, &options);
  options = rb_funcall(self, rb_intern("merge_defaults"), 1, options);
  VALUE user_agent    = hn_hash_lookup_sym(options, "user_agent"),
        settings_path = hn_hash_lookup_sym(options, "settings_path"),
        cache_path    = hn_hash_lookup_sym(options, "cache_path"),
        
        load_playlists = hn_hash_lookup_sym(options, "load_playlists"),
        compress_playlists = hn_hash_lookup_sym(options, "compress_playlists"),
        cache_playlist_metadata = hn_hash_lookup_sym(options, "cache_playlist_metadata");
  
  // user_agent: max 255 characters long
  if (rb_str_strlen(user_agent) > 255)
  {
    rb_raise(rb_eArgError, "User-Agent may not be more than 255 characters");
  }
  
  /*
    Finally, we do the libspotify dance and spawn our threads.
  */
  sp_session_config config =
  {
    .api_version          = SPOTIFY_API_VERSION,
    .cache_location       = StringValueCStr(cache_path),
    .settings_location    = StringValueCStr(settings_path),
    .application_key      = StringValuePtr(appkey),
    .application_key_size = RSTRING_LENINT(appkey),
    .user_agent           = StringValueCStr(user_agent),
    .callbacks            = &HALLON_SESSION_CALLBACKS,
    .userdata             = session_data,
    .compress_playlists   = RTEST(compress_playlists),
    .initially_unload_playlists = ! RTEST(load_playlists),
    .dont_save_metadata_for_playlists = ! RTEST(cache_playlist_metadata),
  };
  
  /* @note This calls the `notify_main_thread` callback once from the same pthread. */
  sp_error error = sp_session_create(&config, (sp_session**) session_data->spotify_ptr);
  hn_eError_maybe_raise(error);
  
  /* spawn the event producer & consumer */
  VALUE threads = rb_funcall2(self, rb_intern("spawn_handlers"), 0, NULL);
  
  rb_iv_set(self, "@threads", threads);
  rb_iv_set(self, "@appkey", appkey);
  rb_iv_set(self, "@options", options);
  
  return self;
}

/*
  Waits for the global event variable to contain an event. Once an event
  arrives, it is sent to the event queue and then handled by its’ handler.
  
  @note We use a separate thread for firing the event to avoid deadlock.
  
  @param [Queue] queue event queue
  @return [Thread] taskmaster thread
*/
static VALUE cSession_spawn_taskmaster(VALUE self, VALUE queue)
{
  return rb_thread_create(taskmaster_thread /* callbacks.c */, (void*) queue);
}

/*
  Retrieve the connection state for this session.
  
  @return [Symbol] `:logged_out`, `:logged_in`, `:disconnected` or `:undefined`
*/
static VALUE cSession_status(VALUE self)
{
  switch(sp_session_connectionstate(SESSPTR_OF(self)))
  {
    case SP_CONNECTION_STATE_LOGGED_OUT: return STR2SYM("logged_out");
    case SP_CONNECTION_STATE_LOGGED_IN: return STR2SYM("logged_in");
    case SP_CONNECTION_STATE_DISCONNECTED: return STR2SYM("disconnected");
    default: return STR2SYM("undefined");
  }
}

/*
  Processes Spotify events using `sp_session_process_events` until the returned
  timeout is > 0.
  
  @private
  @note Callbacks might be invoked as a side-effect of executing this method.
  @return [Fixnum] milliseconds until {#process_events} should be called again
*/
static VALUE cSession_process_events(VALUE self)
{
  int timeout = (int) hn_proc_without_gvl(sp_session_process_events_nogvl, SESSPTR_OF(self));
  return INT2FIX(timeout);
}

/* this call might lead to trying to lock the event lock, so it is a blocking call */
static VALUE sp_session_process_events_nogvl(void *session_ptr)
{
  int timeout = 0;
  while(timeout == 0) sp_session_process_events((sp_session*) session_ptr, &timeout);
  return (VALUE) timeout;
}

/*
  @overload login(username, password)
    Logs in to Spotify using the given account name and password.
  
    @note This call returns immediately. You are not logged in until
          the `logged_in` event has been fired on this Session.
    @param [#to_s] username
    @param [#to_s] password
    @return [Session]
*/
static VALUE cSession_login(VALUE self, VALUE username, VALUE password)
{
  void *argv[] = { SESSPTR_OF(self), StringValueCStr(username), StringValueCStr(password) };
  hn_proc_without_gvl(sp_session_login_nogvl, argv);
  return self;
}

/* just paranoia, actually */
static VALUE sp_session_login_nogvl(void *_argv)
{
  void **argv = (void**) _argv;
  sp_session_login((sp_session*) argv[0], (char*) argv[1], (char*) argv[2]);
  return Qnil;
}

/*
  @overload fire!(receiver, method, *arguments)
    Fires an event, as if it was generated by `libspotify`.
  
    What this does is that it puts an event in the event queue; ready to
    be picked up by the event dispatcher. Calling this method is pretty
    much the same thing as doing the following:
    `receiver.send(:"on_#{method}", *arguments)`
        
    The only difference is that the method called by the same thread
    that dispatches events from libspotify.
  
    @param [Object] receiver
    @param [#to_s] method
    @param [Object, …] arguments
    @return [Session]
*/
static VALUE cSession_fire_bang(int argc, VALUE *argv, VALUE self)
{
  VALUE recv, method, brgs;
  rb_scan_args(argc, argv, "2*", &recv, &method, &brgs);
  void *args[] = { (void*) recv, (void*) rb_ary_unshift(brgs, method) };
  hn_proc_without_gvl(hn_as_callback_fire, args);
  return self;
}

/*
  Logs out of Spotify. Does nothing if not logged in.
  
  @note This call returns immediately. You are not logged out until
        the `logged_out` event has been fired on this session.
  @return [Session]
*/
static VALUE cSession_logout(VALUE self)
{
  if (rb_funcall3(self, rb_intern("logged_in?"), 0, NULL) == Qtrue)
  {
    hn_proc_without_gvl(sp_session_logout_nogvl, SESSPTR_OF(self));
  }
  
  return self;
}

static VALUE sp_session_logout_nogvl(void *session_ptr)
{
  sp_session_logout(session_ptr);
  return Qnil;
}

/*
  The Session is fundamental for all communication with Spotify. Pretty much *all*
  API calls require you to have established a session with Spotify before
  using them.
  
  @see https://developer.spotify.com/en/libspotify/docs/group__session.html
*/
void Init_Session(void)
{
  VALUE cSession = rb_define_class_under(hn_mHallon, "Session", hn_cBase);
  rb_define_method(cSession, "initialize", cSession_initialize, -1);
  rb_define_method(cSession, "status", cSession_status, 0);
  rb_define_method(cSession, "process_events", cSession_process_events, 0);
  rb_define_method(cSession, "login", cSession_login, 2);
  rb_define_method(cSession, "fire!", cSession_fire_bang, -1);
  rb_define_method(cSession, "logout", cSession_logout, 0);
  
  rb_define_private_method(cSession, "spawn_taskmaster", cSession_spawn_taskmaster, 1);
  
  extern hn_event_t * g_event;
  g_event = ALLOC(hn_event_t);
  g_event->sem_empty  = hn_sem_init(1);
  g_event->sem_full   = hn_sem_init(0);
  g_event->rb_handler = Qnil;
  g_event->c_handler  = NULL;
  g_event->c_data     = NULL;
}

File: lib/hallon/session.rb
# coding: utf-8
require 'singleton'
require 'timeout'
require 'thread'
require 'uri'

module Hallon
  # The Session is fundamental for all communication with Spotify.
  # Pretty much all API calls require you to have established a session
  # with Spotify before using them.
  #
  # @see https://developer.spotify.com/en/libspotify/docs/group__session.html
  class Session < Base
    # The options Hallon used at {Session#initialize}.
    #
    # @return [Hash]
    attr_reader :options

    # The current session cache size (in megabytes).
    #
    # @note This is not provided by libspotify, and the value is only valid
    #       as long as the cache size is only adjusted through {#cache_size=}
    #       and not the Spotify FFI interface.
    #
    # @return [Integer]
    attr_reader :cache_size

    # libspotify only allows one session per process.
    include Singleton
    class << self
      undef :instance
    end

    # We have Session callbacks that you can listen to!
    extend Observable::Session

    # Initializes the Spotify session. If you need to access the
    # instance at a later time, you can use {instance}.
    #
    # @see Session.instance
    #
    # @param (see Session#initialize)
    # @option (see Session#initialize)
    # @yield (see Session#initialize)
    # @raise (see Session#initialize)
    # @see (see Session#initialize)
    # @return [Session]
    def Session.initialize(appkey, options = {}, &block)
      raise "Session has already been initialized" if @__instance__
      @__instance__ = new(appkey, options, &block)
    end

    # Returns the previously initialized Session.
    #
    # @see Session.initialize
    #
    # @return [Session]
    def Session.instance
      @__instance__ or raise NoSessionError, "Session has not been initialized"
    end

    # @return [Boolean] true if a Session instance exists.
    def Session.instance?
      !! @__instance__
    end

    # @return [Array<Symbol>] list of available connection types.
    def self.connection_types
      Spotify.enum_type(:connection_type).symbols
    end

    # @return [Array<Symbol>] list of available connection rules.
    def self.connection_rules
      Spotify.enum_type(:connection_rules).symbols
    end

    # Create a new Spotify session.
    #
    # @param [#to_s] appkey
    # @param [Hash] options
    # @option options [String] :user_agent ("Hallon") User-Agent to use (length < `256`)
    # @option options [String] :settings_path ("tmp") where to save settings and user-specific cache
    # @option options [String] :cache_path ("") where to save cache files (`""` to disable)
    # @option options [String] :tracefile (nil) path to libspotify API tracefile (`nil` to disable)
    # @option options [String] :device_id (nil) device ID for offline synchronization (`nil` to disable)
    # @option options [String] :proxy (nil) proxy URI (supports http, https, socks4, socks5)
    # @option options [Bool]   :load_playlists (true) load playlists into RAM on startup
    # @option options [Bool]   :compress_playlists (true) compress local copies of playlists
    # @option options [Bool]   :cache_playlist_metadata (true) cache metadata for playlists locally
    # @yield allows you to define handlers for events (see {Observable#on})
    # @raise [ArgumentError] if `options[:user_agent]` is more than 256 characters long
    # @raise [Hallon::Error] if `sp_session_create` fails
    # @see http://developer.spotify.com/en/libspotify/docs/structsp__session__config.html
    def initialize(appkey, options = {}, &block)
      if options[:proxy]
        proxy_uri = URI(options[:proxy])
        options[:proxy_username] ||= proxy_uri.user
        options[:proxy_password] ||= proxy_uri.password
        proxy_uri.user = proxy_uri.password = nil
        options[:proxy] = proxy_uri.to_s
      end

      @options = {
        :user_agent => "Hallon",
        :settings_path => "tmp/hallon/",
        :cache_path => "tmp/hallon/",
        :load_playlists => true,
        :compress_playlists => true,
        :cache_playlist_metadata => true,
        :device_id => nil,
        :proxy     => nil,
        :proxy_username => nil,
        :proxy_password => nil,
        :tracefile => nil,
      }.merge(options)

      if @options[:user_agent].bytesize > 255
        raise ArgumentError, "User-agent must be less than 256 bytes long"
      end

      # Default cache size is 0 (automatic)
      @cache_size = 0

      subscribe_for_callbacks do |callbacks|
        config = Spotify::SessionConfig.new
        config[:api_version]   = Hallon::API_VERSION
        config.application_key = appkey
        @options.each { |(key, value)| config.send(:"#{key}=", value) }
        config[:callbacks]     = callbacks

        instance_eval(&block) if block_given?

        # You pass a pointer to the session pointer to libspotify >:)
        FFI::MemoryPointer.new(:pointer) do |p|
          Error::maybe_raise Spotify.session_create(config, p)
          @pointer = p.read_pointer
        end
      end
    end

    # Flushes the Session cache to disk.
    #
    # @note libspotify does this automatically periodically, under normal
    #       circumstances this method should not need to be used.
    # @return [Session]
    def flush_caches
      Spotify.session_flush_caches(pointer)
    end

    # PlaylistContainer for the currently logged in session.
    #
    # @note returns nil if the session is not logged in.
    # @return [PlaylistContainer, nil]
    def container
      container = Spotify.session_playlistcontainer!(pointer)
      PlaylistContainer.from(container)
    end

    # Process pending Spotify events (might fire callbacks).
    #
    # @return [Fixnum] time (in milliseconds) until it should be called again
    def process_events
      FFI::MemoryPointer.new(:int) do |p|
        Spotify.session_process_events(pointer, p)
        return p.read_int
      end
    end

    # Log into Spotify using the given credentials.
    #
    # @example logging in with password
    #   session.login 'Kim', 'password'
    #
    # @example logging in with credentials blob
    #   session.login 'Kim', Hallon::Blob('blob string')
    #
    # @note it also supports logging in via a credentials blob, if you pass
    #       a Hallon::Blob(blob_string) as the password instead of the real password
    # @param [String] username
    # @param [String] password_or_blob
    # @param [Boolean] remember_me have libspotify remember credentials for {#relogin}
    # @return [Session]
    # @see login!
    def login(username, password, remember_me = false)
      if username.empty? or password.empty?
        raise ArgumentError, "username and password may not be blank"
      end

      password, blob = blob, password if password.is_a?(Blob)
      tap { Spotify.session_login(pointer, username, password, remember_me, blob) }
    end

    # Login the remembered user (see {#login}).
    #
    # @raise [Spotify::Error] if no credentials are stored in libspotify
    # @see #relogin!
    def relogin
      Spotify.session_relogin!(pointer)
    end

    # Log in to Spotify using the given credentials.
    #
    # @note This function will not return until you’ve either logged in successfully,
    #       or until an error is raised.
    # @param (see #login)
    # @return [Session]
    # @raise [Error] if failed to log in
    # @see #login
    def login!(username, password, remember_me = false)
      login(username, password, remember_me)
      tap { wait_until_logged_in }
    end

    # Log in the remembered user.
    #
    # @note This method will not return until you’ve either logged in successfully
    #       or until an error is raised.
    # @return [Session]
    # @raise [Error] if failed to log in
    # @see #relogin
    def relogin!
      relogin
      tap { wait_until_logged_in }
    end

    # Log out the current user.
    #
    # @note This method will not return until you’ve logged out successfully.
    # @return [Session]
    def logout!
      logout
      tap { wait_for(:logged_out) { logged_out? } }
    end

    # @return [String, nil] username of the currently logged in user.
    def username
      username = Spotify.session_user_name(pointer)
      username unless username.nil? or username.empty?
    end

    # @return [String, nil] username of the user stored in libspotify-remembered credentials.
    def remembered_user
      bufflen = Spotify.session_remembered_user(pointer, nil, 0)
      FFI::Buffer.alloc_out(bufflen + 1) do |b|
        Spotify.session_remembered_user(pointer, b, b.size)
        return b.get_string(0)
      end if bufflen > 0
    end

    # Remove stored login credentials in libspotify.
    #
    # @note If no credentials are stored nothing’ll happen.
    # @return [self]
    def forget_me!
      tap { Spotify.session_forget_me(pointer) }
    end

    # Logs out of Spotify. Does nothing if not logged in.
    #
    # @return [self]
    def logout
      tap { Spotify.session_logout(pointer) if logged_in? }
    end

    # @return [User] the User currently logged in.
    def user
      user = Spotify.session_user!(pointer)
      User.from(user)
    end

    # @return [Symbol] current connection status.
    def status
      Spotify.session_connectionstate(pointer)
    end

    # @return [Boolean] true if the session is currently set to private.
    def private?
      Spotify.session_is_private_session(pointer)
    end
    alias_method :britney_spears_mode?, :private?

    # Set private session.
    #
    # @note mode is reverted to normal after some time without user activity,
    #       see official libspotify documentation for details.
    # @param [Boolean] is_private
    def private=(is_private)
      Spotify.session_set_private_session(pointer, !! is_private)
    end
    alias_method :britney_spears_mode=, :private=

    # Set session cache size in megabytes.
    #
    # @param [Integer]
    # @return [Integer]
    def cache_size=(size)
      Spotify.session_set_cache_size(pointer, @cache_size = size)
    end

    # @return [String] currently logged in users’ country.
    def country
      coded = Spotify.session_user_country(pointer)
      country = ((coded >> 8) & 0xFF).chr
      country << (coded & 0xFF).chr
    end

    # Star the given tracks.
    #
    # @example
    #   track = Hallon::Track.new("spotify:track:2LFQV2u6wXZmmySCWBkYGu")
    #   session.star(track)
    #
    # @note (see #unstar)
    # @raise (see #unstar)
    #
    # @param [Track…]
    # @return [Session]
    def star(*tracks)
      tap { tracks_starred(tracks, true) }
    end

    # Unstar the given tracks.
    #
    # @example
    #   track = Hallon::Track.new("spotify:track:2LFQV2u6wXZmmySCWBkYGu")
    #   session.unstar(track)
    #
    # @note this method might raise a Spotify::Error, however when this might
    #       occur is not documented in libspotify (and I have yet to find any
    #       way to trigger it myself). it’s entirely possible that this method
    #       never returns an error, but we can’t know for sure.
    #
    # @raise [Spotify:Error] if libspotify reports an error (when this happens is unknown and undocumented)
    # @param [Track…]
    # @return [Session]
    def unstar(*tracks)
      tap { tracks_starred(tracks, false) }
    end

    # Set the connection rules for this session.
    #
    # @raise [ArgumentError] if given invalid connection rules
    # @param [Symbol, …] connection_rules
    # @see Session.connection_rules
    def connection_rules=(connection_rules)
      rules = Array(connection_rules).reduce(0) do |mask, rule|
        mask | Spotify.enum_value!(rule, "connection rule")
      end

      Spotify.session_set_connection_rules(pointer, rules)
    end

    # Set the connection type for this session.
    #
    # @raise [ArgumentError] if given invalid connection type
    # @param [Symbol] connection_type
    # @see Session.connection_types
    def connection_type=(connection_type)
      Spotify.session_set_connection_type(pointer, connection_type)
    end

    # Remaining time left you can stay offline before needing to relogin.
    #
    # @return [Integer] offline time left in seconds
    def offline_time_left
      Spotify.offline_time_left(pointer)
    end

    # Offline synchronization status.
    #
    # @return [Hash] sync status (empty hash if not applicable)
    # @see http://developer.spotify.com/en/libspotify/docs/structsp__offline__sync__status.html
    def offline_sync_status
      struct = Spotify::OfflineSyncStatus.new
      if Spotify.offline_sync_get_status(pointer, struct.pointer)
        Hash[struct.members.zip(struct.values)]
      else
        {}
      end
    end

    # @return [Integer] number of playlists marked for offline sync.
    def offline_playlists_count
      Spotify.offline_num_playlists(pointer)
    end

    # @return [Integer] number of offline tracks left to sync for offline mode.
    def offline_tracks_to_sync
      Spotify.offline_tracks_to_sync(pointer)
    end

    # Set preferred offline bitrate.
    #
    # @example setting offline bitrate without resync
    #   session.offline_bitrate = :'96k'
    #
    # @example setting offline bitrate and resync already-synced tracks
    #   session.offline_bitrate = :'96k', true
    #
    # @note under normal circumstances, ArgumentError is the error that will
    #       be raised on an invalid bitrate. However, if Hallon fails the type
    #       checking (for whatever reason), libspotify will itself return an
    #       error as well.
    #
    # @raise [ArgumentError] if given invalid bitrate
    # @raise [Spotify::Error] if libspotify does not accept the given bitrate (see note)
    # @param [Symbol] bitrate
    # @param [Boolean] resync (default: false)
    # @see Player.bitrates
    def offline_bitrate=(bitrate)
      bitrate, resync = Array(bitrate)
      Spotify.session_preferred_offline_bitrate!(pointer, bitrate, !! resync)
    end

    # @note Returns nil when no user is logged in.
    # @return [Playlist, nil] currently logged in user’s starred playlist.
    def starred
      playlist = Spotify.session_starred_create!(pointer)
      Playlist.from(playlist)
    end

    # @note Returns nil when no user is logged in.
    # @return [Playlist, nil] currently logged in user’s inbox playlist.
    def inbox
      playlist = Spotify.session_inbox_create!(pointer)
      Playlist.from(playlist)
    end

    # @see #status
    # @return [Boolean] true if logged in.
    def logged_in?
      status == :logged_in
    end

    # @see #status
    # @return [Boolean] true if logged out.
    def logged_out?
      status == :logged_out
    end

    # @see #status
    # @return [Boolean] true if session has been disconnected.
    def disconnected?
      status == :disconnected
    end

    # @see #status
    # @return [Boolean] true if offline.
    def offline?
      status == :offline
    end

    private
      # Set starred status of given tracks.
      #
      # @raise [Spotify::Error] … maybe, it’s undocumented in libspotify, who knows?
      # @param [Array<Track>] tracks
      # @param [Boolean] starred
      def tracks_starred(tracks, starred)
        FFI::MemoryPointer.new(:pointer, tracks.size) do |ptr|
          ptr.write_array_of_pointer tracks.map(&:pointer)
          Spotify.track_set_starred!(pointer, ptr, tracks.size, starred)
        end
      end

      # Waits until we’re either logged in or a failure occurs.
      #
      # @note You must call {#login} or {#relogin} before you call this method, or
      #       it will hang forever!
      # @see login!
      # @see relogin!
      def wait_until_logged_in
        wait_for(:logged_in, :connection_error) do |event, error|
          Error.maybe_raise(error)
          session.logged_in?
        end
      end
  end
end
